node/src/reactor/joiner.rs

```diff
@@ -472,6 +480,23 @@ impl reactor::Reactor for Reactor {
                             Ok(block_header) => Some(JoinerEvent::FinishedJoining {
                                 block_header: Box::new(block_header),
                             }),
+                            Err(LinearChainSyncError::RetrievedBlockHeaderFromFutureVersion {
+                                current_version,
+                                block_header_with_future_version,
+                            }) => {
+                                let future_version =
+                                    block_header_with_future_version.protocol_version();
+                                info!(%current_version, %future_version, "Restarting for upgrade");
+                                Some(JoinerEvent::Shutdown(ExitCode::Success))
```

Comment:
I must say I'm always buffled by the fact that `Success` indicates a reason to upgrade...

Reply:
Yes, but I'd rather not change that as part of this PR.

Reply:
Oh, definitely. My modus operandi is to complain long enough that I convince enough people until we do it :)